formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -41,12 +51,13 @@ internal open class StreamingJsonDecoder(
             WriteMode.LIST, WriteMode.MAP, WriteMode.POLY_OBJ -> StreamingJsonDecoder(
                 json,
                 newMode,
-                lexer
+                lexer,
+                descriptor
             )
             else -> if (mode == newMode) {
                 this
```

Comment:
This will leave old descriptor inside new structure with same mode. Try to write test with a nested class e.g. `data class Outer(val i: Inner); data class Inner(val s1: String?, val s2: String?)`